pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -296,15 +298,337 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
+				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
+				glog.Info(message)
+				changes.ContainersToStart[index] = message
+			}
+			continue
+		}
+
+		expectedHash := kubecontainer.HashContainer(&container)
+		containerChanged := containerStatus.Hash != 0 && containerStatus.Hash != expectedHash
+		if containerChanged {
+			message := fmt.Sprintf("Pod %q container %q hash changed (%d vs %d), it will be killed and re-created.",
+				pod.Name, container.Name, containerStatus.Hash, expectedHash)
+			glog.Info(message)
+			changes.ContainersToStart[index] = message
+			continue
+		}
+
+		liveness, found := m.livenessManager.Get(containerStatus.ID)
+		if !found || liveness == proberesults.Success {
+			changes.ContainersToKeep[containerStatus.ID] = index
+			continue
+		}
+		if pod.Spec.RestartPolicy != api.RestartPolicyNever {
+			message := fmt.Sprintf("pod %q container %q is unhealthy, it will be killed and re-created.", format.Pod(pod), container.Name)
+			glog.Info(message)
+			changes.ContainersToStart[index] = message
+		}
+	}
+
+	// compute containers that to be killed
+	runningContainerStatues := podStatus.GetRunningContainerStatuses()
+	for _, containerStatus := range runningContainerStatues {
+		if _, keep := changes.ContainersToKeep[containerStatus.ID]; !keep {
+			var podContainer *api.Container
+			var killMessage string
+			for i, c := range pod.Spec.Containers {
+				if c.Name == containerStatus.Name {
+					podContainer = &pod.Spec.Containers[i]
+					killMessage = changes.ContainersToStart[i]
+					break
+				}
+			}
+
+			changes.ContainersToKill[containerStatus.ID] = containerToKillInfo{
+				name:      containerStatus.Name,
+				container: podContainer,
+				message:   killMessage,
+			}
+		}
+	}
+
+	return changes
+}
+
+// SyncPod syncs the running pod into the desired pod by executing following steps:
+//
+//  1. Compute sandbox and container changes.
+//  2. Kill pod sandbox if necessary.
+//  3. Kill any containers that should not be running.
+//  4. Create sandbox if necessary.
+//  5. Create necessary containers
+//
+// TODO: support init containers in SyncPod.
+func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
+	// Step 1: Compute sandbox and container changes.
+	podContainerChanges := m.computePodContainerChanges(pod, podStatus)
+	glog.V(3).Infof("computePodContainerChanges got %+v for pod %q", podContainerChanges, format.Pod(pod))
+	if podContainerChanges.CreateSandbox {
+		ref, err := api.GetReference(pod)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q: '%v'", format.Pod(pod), err)
+		}
+		if podContainerChanges.SandboxID != "" {
+			m.recorder.Eventf(ref, api.EventTypeNormal, "SandboxChanged", "Pod sandbox changed, it will be killed and re-created.")
```

Comment:
nit: Change to pod, because tt other place we just use `Eventf(pod,`. Just nit. :0